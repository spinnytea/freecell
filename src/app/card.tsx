import Image from "next/image";

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];

export type Rank = 'ace' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | 'jack' | 'queen' | 'king';
export const RankList: Rank[] = ['ace', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'jack', 'queen', 'king'];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor(width / ORIG_WIDTH * ORIG_HEIGHT);

export function CardImage({
	width=ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number,
	rank: Rank,
	suit: Suit,
}) {
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={width}
			height={height}
			priority
		/>
	);
}

function getFilename(rank: Rank, suit: Suit) {
	const fancy = (rank === 'jack' || rank === 'queen' || rank === 'king' ? '2' : '');
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}
