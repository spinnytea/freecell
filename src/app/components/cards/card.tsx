import Image from 'next/image';

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor((width / ORIG_WIDTH) * ORIG_HEIGHT);

const FANCY_DECK = true;

export function CardImage({
	width = ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number;
	rank: Rank;
	suit: Suit;
}) {
	// REVIEW do we need `priority` - it's complaining because i'm rendering the whole deck on init
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			priority
		/>
	);
}

// TODO toggle fancy (also ace of spades)
function getFilename(rank: Rank, suit: Suit, useFancyDeck = FANCY_DECK) {
	if (rank === 'joker') {
		if (isRed(suit)) return `/SVG-cards-1.3/red_joker.svg`;
		return `/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `/SVG-cards-1.3/ace_of_spades.svg`;
		return `/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
