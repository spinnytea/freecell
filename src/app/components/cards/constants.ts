export const ORIG_WIDTH = 208;
export const ORIG_HEIGHT = 303;

export const ASSET_FOLDER = process.env.BASE_PATH ?? '';

export type CardFaces = 'SVGCards13' | 'SmolCards';
