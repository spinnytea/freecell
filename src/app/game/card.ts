// XXX (techdebt) duplicate
const BOTTOM_OF_CASCADE = 99;

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];
export const isAdjacent = ({ min, max }: { min: Rank; max: Rank }) =>
	RankList.indexOf(min) === RankList.indexOf(max) - 1;

// REVIEW (techdebt) is "fixture" the right name?
//  - cell -> freecell
//  - foundation -> homecell
//  - cascade -> column
export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

export type MoveSourceType = 'deck' | 'cell' | 'foundation' | 'cascade:single' | 'cascade:sequence';
export type MoveDestinationType = 'cell' | 'foundation' | 'cascade:empty' | 'cascade:sequence';
export const MoveDestinationTypePriorities: {
	[moveSourceType in MoveSourceType]: { [moveDestinationType in MoveDestinationType]: number };
} = {
	// XXX (controls) deck: when will we get to do this?
	'deck': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cell': {
		'cell': 1,
		'foundation': 3,
		'cascade:empty': 2,
		'cascade:sequence': 4,
	},
	// XXX (controls) foundation: down from foundation means "back into play?"
	'foundation': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cascade:single': {
		'cell': 2,
		'foundation': 3,
		'cascade:empty': 1,
		'cascade:sequence': 4,
	},
	'cascade:sequence': {
		'cell': 1,
		'foundation': 2,
		'cascade:empty': 3,
		'cascade:sequence': 4,
	},
};

/**
	REVIEW (techdebt) does AvailableMove need to store moveDestinationType and priority?
	 - moveDestinationType is needed for building the list (like moveSourceType)
	   but do we need to store it?
	 - priority could be build into the sort order of the list (pick the first item),
	   rather than stored with the location (pick the highest number)
	 - it is nice to see in the DebugCursors the different categories
	 	priority: 'high' | 'low';
	---
	 - we have sufficient impl to simplify now
	 - wait for unit testing of DebugCursors
	 - wait for manualtesting of DebugCursors
*/
export interface AvailableMove {
	/** where we could move the card */
	location: CardLocation;
	/** helps us think about priorities / communicate settings */
	moveDestinationType: MoveDestinationType;
	/** if we are going to visualize them debug mode, we need to have it precomputed */
	priority: number;
}

/**
	foundation: h
	cells: a - d
	cascades: 1 - 9, t (1-8, but we allow 9 and 10 columns)

	@see [Standard FreeCell Notation](https://www.solitairelaboratory.com/solutioncatalog.html)
*/
export type Position =
	| 'a'
	| 'b'
	| 'c'
	| 'd'
	| 'h'
	| '1'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| 't';

export interface Card {
	rank: Rank;
	suit: Suit;
	location: CardLocation;
}

export interface CardSequence {
	/**
		the first card in the sequence
		(cursor, selection)
	*/
	location: CardLocation;

	/**
		the list of cards in the sequence
		(alternates red/black, rank descends by one)
	*/
	cards: Card[];

	/** since we are allowing any card to be selected or inspected, not all of them are movable */
	canMove: boolean;
}

export function isLocationEqual(a: CardLocation, b: CardLocation) {
	return a.fixture === b.fixture && a.data[0] === b.data[0] && a.data[1] === b.data[1];
}

export function shorthandCard(card: Card | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank === 'joker' ? 'W' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}

export function parseShorthandCard(
	r: string | undefined,
	s: string | undefined
): { rank: Rank; suit: Suit } | null {
	if (r === ' ' && s === ' ') return null;

	let rank: Rank;
	let suit: Suit;
	switch (r) {
		case 'A':
			rank = 'ace';
			break;
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			rank = r;
			break;
		case 'T':
			rank = '10';
			break;
		case 'J':
			rank = 'jack';
			break;
		case 'Q':
			rank = 'queen';
			break;
		case 'K':
			rank = 'king';
			break;
		case 'W':
			rank = 'joker';
			break;
		default:
			throw new Error(`invalid rank shorthand: "${r ?? 'undefined'}"`);
	}

	switch (s) {
		case 'C':
			suit = 'clubs';
			break;
		case 'D':
			suit = 'diamonds';
			break;
		case 'H':
			suit = 'hearts';
			break;
		case 'S':
			suit = 'spades';
			break;
		default:
			throw new Error(`invalid suit shorthand: "${s ?? 'undefined'}"`);
	}

	return { rank, suit };
}

export function shorthandSequence(sequence: CardSequence, includePosition = false) {
	const cards = sequence.cards.map((card) => shorthandCard(card)).join('-');

	if (sequence.canMove && includePosition) {
		return shorthandPosition(sequence.location) + ' ' + cards;
	}

	return cards;
}

export function shorthandPosition(location: CardLocation): Position {
	const d0 = location.data[0];
	if (location.fixture === 'foundation') {
		return 'h';
	} else if (location.fixture === 'cascade') {
		// this doesn't check data[1], it assumes it's the final row
		// sequences would need to check data[1] + card.length, not just the location
		// (could pass in a optional canMove with default of true)
		if (d0 === 9) {
			return 't';
		} else if (d0 >= 0 && d0 < 9) {
			return (d0 + 1).toString(10) as Position;
		}
	} else if (location.fixture === 'cell') {
		if (d0 >= 0 && d0 < 6) {
			return (d0 + 10).toString(16) as Position;
		}
	}
	throw new Error(`invalid position: ${JSON.stringify(location)}`);
}

/**
	this is part 1 of a 2 step process
	moves are not always obvious
	i.e. 1h - _which_ foundation do we use? there are 4
	i.e. 42 - is this a cascade:single or cascade:sequence moving to a cascade:sequence or cascade:empty?

	notice also that this function only accepts the single character, it does not accept a game
	so which d1 do we use for a cascade? this will return an invalid value (too high), which will be clamped if used directly
*/
export function parseShorthandPosition_INCOMPLETE(p: string | undefined): CardLocation {
	switch (p) {
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			// this isn't a valid cursor position, it will need to be clamped
			// cascades can have sequences, so you need to decide if you really want the "bottom"
			return { fixture: 'cascade', data: [parseInt(p, 10) - 1, BOTTOM_OF_CASCADE] };
		case 't':
			return { fixture: 'cascade', data: [9, BOTTOM_OF_CASCADE] };
		case 'h':
			// h could refer to _any_ of the foundations; this needs to be verified
			return { fixture: 'foundation', data: [0] };
		case 'a':
		case 'b':
		case 'c':
		case 'd':
		case 'e':
		case 'f':
			return { fixture: 'cell', data: [parseInt(p, 16) - 10] };
		default:
			throw new Error(`invalid position shorthand: "${p ?? 'undefined'}"`);
	}
}
