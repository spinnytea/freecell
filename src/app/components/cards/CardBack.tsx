import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';

// TODO (theme) alternate card backs?
// TODO (hud) (2-priority) Airplane mode: no deck?
// - Does it check for it every time?
// - Do we need to place an invisible one / off screen so it's always loaded?
export function CardBack({
	width,
	height,
}: Readonly<{
	width: number;
	height: number;
}>) {
	const filename = getFilename();
	return (
		<Image
			src={filename}
			alt="card back"
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function getFilename() {
	return `${ASSET_FOLDER}/i/Card_back_10.svg`;
}
