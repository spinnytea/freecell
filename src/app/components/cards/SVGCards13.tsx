import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';
import { isRed, Rank, Suit } from '@/app/game/card';

export function SVGCards13({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	const filename = getFilename(rank, suit);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function getFilename(rank: Rank, suit: Suit, useFancyDeck = true) {
	if (rank === 'joker') {
		if (isRed(suit)) return `${ASSET_FOLDER}/i/SVG-cards-1.3/red_joker.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `${ASSET_FOLDER}/i/SVG-cards-1.3/ace_of_spades.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `${ASSET_FOLDER}/i/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
