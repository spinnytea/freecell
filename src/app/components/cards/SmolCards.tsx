import { ReactElement } from 'react';
import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';
import { ace_of_clubs, ace_of_clubs_defs } from '@/app/components/cards/smol/ace_of_clubs';
import { ace_of_diamonds, ace_of_diamonds_defs } from '@/app/components/cards/smol/ace_of_diamonds';
import { ace_of_hearts, ace_of_hearts_defs } from '@/app/components/cards/smol/ace_of_hearts';
import { ace_of_spades, ace_of_spades_defs } from '@/app/components/cards/smol/ace_of_spades';
import { clubs } from '@/app/components/cards/smol/clubs';
import { diamonds } from '@/app/components/cards/smol/diamonds';
import { hearts } from '@/app/components/cards/smol/hearts';
import { spades } from '@/app/components/cards/smol/spades';
import { isRed, Rank, Suit } from '@/game/card/card';

const SYMBOLS: { [suit in Suit]: ReactElement } = {
	clubs,
	diamonds,
	hearts,
	spades,
};
const ACE_SYMBOL_DEFS: { [suit in Suit]: ReactElement } = {
	clubs: ace_of_clubs_defs,
	diamonds: ace_of_diamonds_defs,
	hearts: ace_of_hearts_defs,
	spades: ace_of_spades_defs,
};
const ACE_SYMBOLS: { [suit in Suit]: ReactElement } = {
	clubs: ace_of_clubs,
	diamonds: ace_of_diamonds,
	hearts: ace_of_hearts,
	spades: ace_of_spades,
};
const TEXT: { [rank in Rank]: string } = {
	'ace': 'A',
	'2': '2',
	'3': '3',
	'4': '4',
	'5': '5',
	'6': '6',
	'7': '7',
	'8': '8',
	'9': '9',
	'10': '10',
	'jack': 'J',
	'queen': 'Q',
	'king': 'K',
	'joker': 'W', // unused
};

export function SmolCards({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	if (rank !== 'joker') {
		return <NativeCards width={width} height={height} rank={rank} suit={suit} />;
	}

	// XXX (accessibility) consider adding title={rank or suit}
	const filename = getFilename(rank, suit);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function NativeCards({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	const color = isRed(suit) ? '#df0000' : '#000000';
	const text = TEXT[rank];
	const symbol = SYMBOLS[suit];

	// XXX (accessibility) consider adding <title>rank or suit</title>
	return (
		<svg
			width={width}
			height={height}
			viewBox="0 0 208 303"
			xmlns="http://www.w3.org/2000/svg"
			xmlnsXlink="http://www.w3.org/1999/xlink"
		>
			<desc>
				{rank} of {suit}
			</desc>

			{rank === 'ace' && ACE_SYMBOL_DEFS[suit]}

			<rect width="208" height="303" rx="8" fill="#FFFFFF" stroke="#000000" strokeWidth="0.5" />

			<g fill={color} stroke="none" textAnchor="middle" dominantBaseline="middle" fontWeight={500}>
				<text fontSize="90" x="52" y="56" letterSpacing="-0.1em">
					{text}
				</text>
			</g>

			<g>{symbol}</g>

			{rank === 'ace' ? (
				<g transform="scale(1.2) translate(-19, -40)">{ACE_SYMBOLS[suit]}</g>
			) : (
				<g transform="scale(2) translate(-95, 42)">{symbol}</g>
			)}
		</svg>
	);
}

function getFilename(rank: Rank, suit: Suit) {
	if (rank === 'joker') {
		// TODO (joker) restyle joker
		//  - (make one for each suit, rotate the center so suit is top-left)
		if (isRed(suit)) return `${ASSET_FOLDER}/i/SVG-cards-1.3/red_joker.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/black_joker.svg`;
	}

	return `${ASSET_FOLDER}/i/smol-cards/ace_of_${suit}.svg`;
}
