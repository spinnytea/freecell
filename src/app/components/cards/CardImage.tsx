import { CardBack } from '@/app/components/cards/CardBack';
import {
	CARD_FACE_CUTOFF,
	CardFaces,
	ORIG_WIDTH,
	scale_height,
} from '@/app/components/cards/constants';
import { SmolCards } from '@/app/components/cards/SmolCards';
import { SVGCards13 } from '@/app/components/cards/SVGCards13';
import { Rank, Suit } from '@/app/game/card/card';
import { useSettings } from '@/app/hooks/contexts/Settings/useSettings';

// TODO (theme) https://cardmeister.github.io/
//  - https://github.com/cardmeister/cardmeister.github.io/blob/master/elements.cardmeister.full.js
// TODO (theme) https://en.wikipedia.org/wiki/Playing_cards_in_Unicode
// TODO (theme) alternate card backs?
// TODO (theme) deck w/ kings are lions?
// TODO (theme) dark theme cards
// TODO (offline) if SVGCards13 is unavailable, fallback to SmolCards
export function CardImage({
	hidden = false,
	rank,
	suit,
	width = ORIG_WIDTH,
}: Readonly<{
	hidden?: boolean;
	rank: Rank;
	suit: Suit;
	width?: number;
}>) {
	const { cardFace } = useSettings();
	const faces: CardFaces =
		cardFace === 'auto' ? (width < CARD_FACE_CUTOFF ? 'SmolCards' : 'SVGCards13') : cardFace;
	const height = scale_height(width);

	if (hidden) {
		return <CardBack width={width} height={height} />;
	}

	switch (faces) {
		case 'SVGCards13':
		default:
			return <SVGCards13 width={width} height={height} rank={rank} suit={suit} />;
		case 'SmolCards':
			return <SmolCards width={width} height={height} rank={rank} suit={suit} />;
	}
}
