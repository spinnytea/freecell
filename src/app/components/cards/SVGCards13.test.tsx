import { units } from '@/app/components/cards/SVGCards13';
import { RankList, SuitList } from '@/game/card/card';

describe('SVGCards13', () => {
	test('getFilename', () => {
		const allNames: string[] = [];

		SuitList.forEach((suit) => {
			RankList.forEach((rank) => {
				allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit)}`);
			});
		});
		SuitList.forEach((suit) => {
			const rank = 'joker';
			allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit)}`);
		});
		// allNames.push(`card back → ${units.getFilename('ace', 'spades')}`);

		expect(allNames).toMatchSnapshot();
	});
});
