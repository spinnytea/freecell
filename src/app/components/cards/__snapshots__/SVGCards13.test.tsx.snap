// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGCards13 getFilename 1`] = `
[
  "ace of clubs → ace_of_clubs.svg",
  "2 of clubs → 2_of_clubs.svg",
  "3 of clubs → 3_of_clubs.svg",
  "4 of clubs → 4_of_clubs.svg",
  "5 of clubs → 5_of_clubs.svg",
  "6 of clubs → 6_of_clubs.svg",
  "7 of clubs → 7_of_clubs.svg",
  "8 of clubs → 8_of_clubs.svg",
  "9 of clubs → 9_of_clubs.svg",
  "10 of clubs → 10_of_clubs.svg",
  "jack of clubs → jack_of_clubs2.svg",
  "queen of clubs → queen_of_clubs2.svg",
  "king of clubs → king_of_clubs2.svg",
  "ace of diamonds → ace_of_diamonds.svg",
  "2 of diamonds → 2_of_diamonds.svg",
  "3 of diamonds → 3_of_diamonds.svg",
  "4 of diamonds → 4_of_diamonds.svg",
  "5 of diamonds → 5_of_diamonds.svg",
  "6 of diamonds → 6_of_diamonds.svg",
  "7 of diamonds → 7_of_diamonds.svg",
  "8 of diamonds → 8_of_diamonds.svg",
  "9 of diamonds → 9_of_diamonds.svg",
  "10 of diamonds → 10_of_diamonds.svg",
  "jack of diamonds → jack_of_diamonds2.svg",
  "queen of diamonds → queen_of_diamonds2.svg",
  "king of diamonds → king_of_diamonds2.svg",
  "ace of hearts → ace_of_hearts.svg",
  "2 of hearts → 2_of_hearts.svg",
  "3 of hearts → 3_of_hearts.svg",
  "4 of hearts → 4_of_hearts.svg",
  "5 of hearts → 5_of_hearts.svg",
  "6 of hearts → 6_of_hearts.svg",
  "7 of hearts → 7_of_hearts.svg",
  "8 of hearts → 8_of_hearts.svg",
  "9 of hearts → 9_of_hearts.svg",
  "10 of hearts → 10_of_hearts.svg",
  "jack of hearts → jack_of_hearts2.svg",
  "queen of hearts → queen_of_hearts2.svg",
  "king of hearts → king_of_hearts2.svg",
  "ace of spades → ace_of_spades.svg",
  "2 of spades → 2_of_spades.svg",
  "3 of spades → 3_of_spades.svg",
  "4 of spades → 4_of_spades.svg",
  "5 of spades → 5_of_spades.svg",
  "6 of spades → 6_of_spades.svg",
  "7 of spades → 7_of_spades.svg",
  "8 of spades → 8_of_spades.svg",
  "9 of spades → 9_of_spades.svg",
  "10 of spades → 10_of_spades.svg",
  "jack of spades → jack_of_spades2.svg",
  "queen of spades → queen_of_spades2.svg",
  "king of spades → king_of_spades2.svg",
  "joker of clubs → black_joker.svg",
  "joker of diamonds → red_joker.svg",
  "joker of hearts → red_joker.svg",
  "joker of spades → black_joker.svg",
]
`;
